### ISSUE TITLE
Testing for approximate equality of floats.

### ISSUE THREAD

**Issue by jacg** [iliekturtles/uom#288]
How can one write tests which check for approximate equality of `uom` floats?

There are `uom::tests::Test::{assert_approx_eq, approx_eq)` but `uom::tests` is private, so it's not available to `uom` users.

External crates such as [`float-cmp`](https://crates.io/crates/float-cmp), [`assert_approx_eq`](https://crates.io/crates/assert_approx_eq) or [`float_eq`](https://crates.io/crates/float_eq) all give type errors when used with `uom` types.

**Comment by iliekturtles** [comment]
The simplest way would be to extract a unit value as the underlying storage type. `uom`'s test code is essentially doing this. e.g. for `length`:

```rust
assert_approx_eq!(lhs.get<meter>(), rhs.get<meter>())
```

Do you have a preferred approximate equality crate that you would prefer to use? I can take a look at their macros to see if anything else is possible.

**Comment by jacg** [comment]
I don't have a preference. I've been using `assert_approx_eq` but can't remember why I chose it. It was probably the first one I tried and it let me get on with my life. `float_eq` seems rather well documented, having a book which looks quite thorough at first glance. `float-cmp` looks to be similar to `float_eq`, both having more features than `assert_approx_eq` which only seems to support absolute tolerances.

So I suspect I'll be trying `float-cmp` at sometime in the future, but haven't done so yet.

**Comment by jacg** [comment]
I've taken [`float_eq`](https://crates.io/crates/float_eq) for a quick spin. First impressions very good. I'll probably use it instead of `assert_approx_eq` from now on.

Following your suggestion, this macro

```rust
 macro_rules! assert_uom_eq {
        ($unit:ident, $lhs:expr, $rhs:expr, $algo:ident <= $tol:expr) => {
            assert_float_eq!($lhs.get::<$unit>(), $rhs.get::<$unit>(), $algo <= $tol)
        };
    }
```
makes using `float_eq` with `uom` quite pleasant.

**Comment by adamreichold** [comment]
> Thoughts?

I think the established procedure is to expose these trait implementations as optional dependencies but this can get messy, e.g. supporting multiple incompatible versions or alternative crates, c.f. https://docs.rs/ndarray/latest/ndarray/#crate-feature-flags

That said, I suspect that `approx` has a bit more clout in the ecosystem at the moment and since it is already used internally, exposing it as an optional public dependency seems like the easiest path forward.

Admittedly, I did not look at `float_eq`, so I do not know whether it brings anything to the table that is really fundamentally better than `approx`. In any case, one could support both `approx` and `float_eq` as optional dependencies at the same time at the cost of additional maintenance effort.

**Comment by iliekturtles** [comment]
I didn't review `float_eq` closely either so I can't say how favorably it compares to the other crates or how they compare against each other. My main point of hesitation is the maintenance burden and requests to support multiple approximate equality crates. Right now `approx` has the most overall downloads but [`float-cmp`](https://crates.io/crates/float-cmp) has higher recent downloads. Both crates have significantly more usage than `float_eq`. Do we pick one. If more than one what is the cutoff for how many of these crates are supported?

Users being unable to implement the traits in their own code is a big downside.

**Comment by jacg** [comment]
I don't think that these crates are particularly complex or heavy, or that the differences between the featureful ones are all that significant. So I'd say it would be reasonable to pick just one, as long as it's not too spartan.

I happen to be happy with `float_eq` today, but if you decide to support some other single crate instead, it won't be a problem.

Though I would be slightly sad to lose the informative messages that `float_eq` generates out of the box:

```
thread 'main' panicked at 'assertion failed: `float_eq!(left, right, r2nd <= t)`
        left: `9.99999999999998`,
       right: `10.0`,
    abs_diff: `0.000000000000019539925233402755`,
   ulps_diff: `Some(11)`,
    [r2nd] t: `0.000000000000002220446049250313`', src\main.rs:15:9
```

Maybe the other crates generate as good, or even better, messages. If they don't, and I really miss them, then it's almost trivial to hack together something like the macro I showed upthread, to let me use my favourite crate.

But I do think it's important that *some* float comparison be available in `uom` out of the box.

(The other advantage that `float_eq` has over the other crates, that I can see in my very superficial review, is that it is documented not only in the usual auto-generate API docs, but also provides a [small book](https://jtempest.github.io/float_eq-rs/book/).)
